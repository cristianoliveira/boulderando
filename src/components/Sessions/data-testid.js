// eslint-disable-next-line
export const SESSION_LIST_TABLE_ITEM = 'session_list_table_item'
