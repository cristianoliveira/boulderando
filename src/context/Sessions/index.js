import { createContext, useContext, useState, useEffect } from 'react'
import { useRouter } from 'next/router'

import useListStorage from '../../hooks/useListStorage'

import { getSessions, postSessionSchedule } from '../../api/bouldering-sessions'

export const SessionContext = createContext()

export const SessionConsumer = SessionContext.Consumer

export function SessionProvider({ children }) {
  const [hasSubmitted, setHasSubmitted] = useState(false)
  const [isProcessing, setIsProcessing] = useState(false)
  const [sessions, setSessions, resetSessions] = useListStorage('sessions', [])
  const router = useRouter()

  const scheduleSession = (session, user) => {
    setHasSubmitted(true)
    setIsProcessing(true)
    return postSessionSchedule(user, session).then(({ data, error }) => {
      setIsProcessing(false)

      if (data) {
        return { data: { ...data, ...session } };
      }

      return { error }
    })
  }

  const addCustomSession = () => {
    router.push('/sessions/new')
  }

  const deleteCustomSessions = () => {
    resetSessions()
  }

  const saveCustomSession = (session) => {
    setSessions([...sessions, session])
    router.push('/sessions')
  }

  useEffect(() => {
    if (!sessions.length) {
      getSessions().then((defaultSessions) => {
        setSessions(defaultSessions)
      })
    }
  }, [])

  return (
    <SessionContext.Provider
      value={{
        sessions,
        addCustomSession,
        deleteCustomSessions,
        scheduleSession,
        saveCustomSession,
        isProcessing,
        hasSubmitted,
      }}
    >
      {children}
    </SessionContext.Provider>
  )
}

export default () => useContext(SessionContext)
