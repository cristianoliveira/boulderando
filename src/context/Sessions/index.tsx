import { createContext, useContext, useState, useEffect } from 'react'
import { useRouter } from 'next/router'

import useListStorage from '../../hooks/useListStorage'
import { SESSIONS } from '../../storage/items'

type SessionContextValue = {
  sessions: StorageData<Session>
  deleteCustomSessions(): void
  scheduleSession(
    session: Session,
    user: User
  ): Promise<{ data?: object; error?: object }>
  saveCustomSession(session: Session): void
  isProcessing: boolean
  hasSubmitted: boolean
} | null

export const SessionContext = createContext<SessionContextValue>(null)

export const SessionConsumer = SessionContext.Consumer

export function SessionProvider({ children, api }: WithChildren & WithApi) {
  const [hasSubmitted, setHasSubmitted] = useState(false)
  const [isProcessing, setIsProcessing] = useState(false)
  const [sessions, setSessions, resetSessions] = useListStorage<Session>(
    SESSIONS,
    []
  )
  const router = useRouter()

  const scheduleSession = (session: Session, user: User) => {
    setHasSubmitted(true)
    setIsProcessing(true)
    return api
      .postSessionSchedule(user, session)
      .then(({ data, error }: any) => {
        setIsProcessing(false)

        if (data) {
          return { data: { ...data, ...session } }
        }

        return { error }
      })
  }

  const deleteCustomSessions = () => {
    resetSessions()
  }

  const saveCustomSession = (session: Session) => {
    setSessions([...sessions, session])
    router.push('/sessions')
  }

  useEffect(() => {
    if (!sessions.length) {
      api.getSessions().then((defaultSessions: Session[]) => {
        setSessions(defaultSessions)
      })
    }
  }, [])

  return (
    <SessionContext.Provider
      value={{
        sessions,
        deleteCustomSessions,
        scheduleSession,
        saveCustomSession,
        isProcessing,
        hasSubmitted,
      }}
    >
      {children}
    </SessionContext.Provider>
  )
}

export default () => useContext(SessionContext)
