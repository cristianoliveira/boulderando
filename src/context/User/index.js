import { createContext, useState, useEffect } from 'react'
import { useRouter } from 'next/router'

import * as storage from '../../storage/local'

const UserContext = createContext()

export const UserConsumer = UserContext.Consumer

export function UserProvider({ guard, children }) {
  const router = useRouter()
  const [user, setUser] = useState(storage.getUser())

  const editUser = () => {
    router.push('/user/edit')
  }

  const deleteUser = () => {
    storage.deleteUser()
    setUser(null)
    router.push('/')
  }

  const saveUser = (user) => {
    storage.saveUser(user)
    setUser(user)
    router.push('/sessions')
  }

  useEffect(() => {
    if (!storage.getUser()) {
      router.push('/')
    }
  }, [])

  return (
    <UserContext.Provider value={{ user, saveUser, deleteUser, editUser }}>
      {children}
    </UserContext.Provider>
  )
}
