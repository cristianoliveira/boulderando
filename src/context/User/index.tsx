import { createContext, useContext, useEffect } from 'react'
import { useRouter } from 'next/router'

import useStorage from '../../hooks/useStorage'
import { USER } from '../../storage/items'
import { getRedirectParam, browserRedirectTo } from '../../modules/redirect-to'

type UserContextValue = {
  user: StorageData<User>
  saveUser(user: User): void
  deleteUser(): void
  editUser(): void
}

export const UserContext = createContext<UserContextValue>(undefined as any)

export const UserConsumer = UserContext.Consumer

const URI_USER_NEW = '/user/new'
const URI_USER_EDIT = '/user/edit'

export function UserProvider({
  children,
  api,
}: WithChildren & WithApi): JSX.Element {
  const router = useRouter()
  const [user, setUser, removeUser] = useStorage<User>(USER)

  const editUser = () => {
    router.push(URI_USER_EDIT)
  }

  const deleteUser = () => {
    removeUser()
    router.push(URI_USER_NEW)
  }

  const saveUser = (userToSave: User) => {
    api.postCreateUser(userToSave).then(() => {
      const redirectTo = getRedirectParam()
      if (redirectTo) {
        browserRedirectTo(redirectTo)
      }

      setUser(userToSave)
      router.push('/')
    })
  }

  useEffect(() => {
    if (
      !user &&
      !router.pathname.includes('sync') &&
      !router.pathname.includes(URI_USER_NEW)
    ) {
      router.push(URI_USER_NEW)
    }
  }, [user])

  return (
    <UserContext.Provider value={{ user, saveUser, deleteUser, editUser }}>
      {children}
    </UserContext.Provider>
  )
}

export default () => useContext<UserContextValue>(UserContext)
