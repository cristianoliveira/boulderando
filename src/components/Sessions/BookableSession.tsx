import React from 'react'
import { TableCell, TableCellProps, TableRow, Button } from '@mui/material'
import EventAvailableIcon from '@mui/icons-material/EventAvailable'

import getNextPossibleDay from '../../modules/weekday'

import { SESSION_LIST_TABLE_ITEM } from './data-testid'

const TbCell = (props: TableCellProps) => (
  <TableCell
    sx={{
      padding: '8px',
    }}
    {...props}
  />
)

type BookableSessionProps = {
  session: Session
  hasBeenBooked: boolean
  isProcessing: boolean
  onBookingSubmit(sessions: Session): void
}

const BookableSession: React.FC<BookableSessionProps> = ({
  session,
  hasBeenBooked,
  isProcessing,
  onBookingSubmit,
}) => (
  <TableRow
    data-testid={SESSION_LIST_TABLE_ITEM}
    sx={{
      color: hasBeenBooked ? 'gray' : 'green',
    }}
  >
    <TbCell>{session.gym_name}</TbCell>
    <TbCell>
      {session.day_of_week}-{getNextPossibleDay(session.day_of_week)}
    </TbCell>
    <TbCell>{session.time}</TbCell>
    <TbCell align={'right'}>
      <Button
        disabled={isProcessing || hasBeenBooked}
        variant="contained"
        type="submit"
        onClick={() => onBookingSubmit(session)}
      >
        <EventAvailableIcon />
      </Button>
    </TbCell>
  </TableRow>
)

export default BookableSession
