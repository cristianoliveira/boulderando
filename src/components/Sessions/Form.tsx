import styled from '@emotion/styled'

import FormControl from '@mui/material/FormControl'
import FormHelperText from '@mui/material/FormHelperText'
import Button from '@mui/material/Button'
import Grid from '@mui/material/Grid'
import { useForm } from 'react-hook-form'
import Alert from '@mui/material/Alert'

import SelectField from '../fields/SelectField'

import * as TID from '../../constants/data-testid'

const StyledForm = styled.form`
  padding: 20px;
`

const FULL_GRID = 12

const showErrorFields = (errors: object): string =>
  Object.keys(errors)
    .map((e) => e)
    .join(', ')

// Sometime in life you have to just have faith
// range from 8:00 to 21:30
const timesSelection: Array<string> = [...Array(20).keys()]
  .filter((i) => i >= 8 && i <= 20)
  .map((i) => [`${i}:00 - ${i + 2}:00`, `${i}:30 - ${i + 2}:30`])
  .flat()

function Form({ onSubmit }: Submitable): JSX.Element {
  const {
    register,
    handleSubmit,
    formState: { errors, isValid, isSubmitted },
  } = useForm({})

  const errorsList = showErrorFields(errors)
  return (
    <StyledForm
      onSubmit={handleSubmit((data) =>
        onSubmit({ ...data, day_of_week: data.human_date.replace(/this /, '') })
      )}
    >
      <Grid
        direction="row"
        alignItems="flex-start"
        spacing={2}
        container
      >
        <Grid item xs={FULL_GRID}>
          <SelectField
            data-testid={TID.CUSTOM_SESSION_FORM_GYM}
            name="gym_name"
            label="Gym"
            register={register}
            selectOptions={[
              { label: 'Boulderklub', value: 'boulderklub' },
              { label: 'Basement', value: 'basement' },
            ]}
          />
        </Grid>
        <Grid item xs={FULL_GRID}>
          <SelectField
            data-testid={TID.CUSTOM_SESSION_FORM_DAY}
            name="human_date"
            label="Day of the week"
            register={register}
            selectOptions={[
              { label: 'Monday', value: 'this monday' },
              { label: 'Tuesday', value: 'this tuesday' },
              { label: 'Wednesday', value: 'this wednesday' },
              { label: 'Thursday', value: 'this thursday' },
              { label: 'Friday', value: 'this friday' },
              { label: 'Saturday', value: 'this saturday' },
            ]}
          />
        </Grid>
        <Grid item xs={FULL_GRID}>
          <SelectField
            data-testid={TID.CUSTOM_SESSION_FORM_TIME}
            name="time"
            label="Time"
            register={register}
            selectOptions={timesSelection.map((time) => ({
              label: time,
              value: time,
            }))}
          />
        </Grid>
        <Grid item xs={FULL_GRID}>
          {isSubmitted && !isValid && errorsList && (
            <Alert
              data-testid={TID.USER_FORM_ERROR_MESSAGE_CONTAINER}
              severity="error"
            >
              All fields are required. Missing fields: {errorsList}
            </Alert>
          )}
        </Grid>
      </Grid>
      <FormControl fullWidth>
        <Button
          data-testid={TID.CUSTOM_SESSION_FORM_SUBMIT_BUTTON}
          variant="contained"
          type="submit"
        >
          Save data
        </Button>
      </FormControl>
      <FormHelperText variant="outlined">
        ** This data is stored in your browser and can be deleted at any time
      </FormHelperText>
    </StyledForm>
  )
}

export default Form
