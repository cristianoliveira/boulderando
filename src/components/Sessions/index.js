import React from 'react'
import { SessionProvider, SessionConsumer } from '../../context/Sessions'
import SessionList from './SessionList'

function Session({ user }) {
  return (
    <SessionProvider user={user}>
      <SessionConsumer>
        {(context) => (
          <SessionList {...context} />
        )}
      </SessionConsumer>
    </SessionProvider>
  )
}

export default Session
