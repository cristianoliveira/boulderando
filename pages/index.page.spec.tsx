import React from 'react'
// eslint-disable-next-line
import renderer from 'react-test-renderer'

import { SessionProvider } from '../src/context/Sessions'
import { UserProvider } from '../src/context/User'
import IndexPage from './index.page'

describe('Form', () => {
  it('matches the snapshot', async () => {
    expect(
      renderer
        .create(
          <UserProvider api={{} as any}>
            <SessionProvider api={{} as any}>
              <IndexPage setPageTitle={jest.fn()} />
            </SessionProvider>
          </UserProvider>
        )
        .toJSON()
    ).toMatchSnapshot()
  })
})
