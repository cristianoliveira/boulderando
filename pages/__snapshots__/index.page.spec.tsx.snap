// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches the snapshot 1`] = `
<div
  className="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
>
  <table
    className="MuiTable-root css-rqglhn-MuiTable-root"
    role={null}
  >
    <thead
      className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1lxfooz-MuiTableCell-root"
          scope="col"
        >
          Gym
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1lxfooz-MuiTableCell-root"
          scope="col"
        >
          Day
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1lxfooz-MuiTableCell-root"
          scope="col"
        >
          Time
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-12l4dq2-MuiTableCell-root"
          scope="col"
        >
          Schedule
        </th>
      </tr>
    </thead>
    <tr
      className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
      role={null}
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-sizeMedium css-bsph2u-MuiTableCell-root"
        colSpan="6"
      />
    </tr>
  </table>
</div>
`;
